pkg/controller/endpoint/endpoints_controller.go

```diff
@@ -555,6 +561,9 @@ func (e *Controller) syncService(ctx context.Context, key string) error {
 
 		return err
 	}
+	if !createEndpoints {
+		e.staleEndpointsTracker.Stale(currentEndpoints)
```

Comment:
we need to store the version of the endpoint we just created or we have updated , lines 535 to 541 of the new change

```
	if createEndpoints {
		// No previous endpoints, create them
		_, err = e.client.CoreV1().Endpoints(service.Namespace).Create(ctx, newEndpoints, metav1.CreateOptions{})
	} else {
		// Pre-existing
		_, err = e.client.CoreV1().Endpoints(service.Namespace).Update(ctx, newEndpoints, metav1.UpdateOptions{})
	}
```

so get the newEndpoint object and store the resource version here in the tracked independently if we created or updated, as we don't want to process it again

Reply:
At least this how I read it in the endpointslice code, but I wonder now if we should return error if is stale and retrying

Reply:
I tried tracking the new endpoints's resource version, but it does resolve the problem or would introduce new problem:
1. If we skip processing a service when the received endpoints update matches the tracked one, it doesn't resolve the original problem because the real problem is that when the controller processes the 2nd pod update (from not ready to ready), it used the stale ready endpoints and skipped updating the endpoints to be ready. Skipping the endpoints event caused by the 1st process changes nothing.
2. If we enqueue the service anyway if the received endpoints update matches the tracked one, then the tracked version would be used to check if the current endpoints match it and the worker would return error if they don't match. But that would cause that once the endpoints is updated by others (user or another controller), this controller will never process it again as the resource version will never match again.

Unlike endpointslice controller, which processes a Service when the endpointslices are equal to or newer than the tracked ones, there is no generation in endpoints object, so we don't know whether an unmatched endpoints is newer or older than the tracked one.

Therefore, instead of tracking the new endpoints's resource version, the patch trackes the old endpoints's resource version. In `syncService`, if it sees the resource version again, it means the cache is out-of-date, and there is no point to proceed. If it sees a different resource version, it means either it has received the last update made by itself, or the endpoints has been updated by others, in either case it should process it.

Reply:
now I got it, thanks